osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/mapping/DataObjectToEventListConverter.java

```diff
@@ -64,32 +63,68 @@ private EventDto getEvent(final DataObject eventDataObject, final EventLogCatego
                     "eventData size should be " + eventLogCategory.getNumberOfEventElements());
         }
 
+        if (eventLogCategory == EventLogCategoryDto.POWER_FAILURE_EVENT_LOG) {
+            return getPowerFailureEvent(eventData, eventLogCategory);
+        } else {
+            return getDefaultEvent(eventData, eventLogCategory);
+        }
+    }
+
+    private EventDto getDefaultEvent(List<DataObject> eventData, final EventLogCategoryDto eventLogCategory)
+            throws ProtocolAdapterException {
         // extract values from List<DataObject> eventData.
         final DateTime dateTime = this.extractDateTime(eventData);
         final Short code = this.extractCode(eventData);
         final Integer eventCounter = this.extractEventCounter(eventLogCategory, eventData);
         final String eventLogCategoryName = eventLogCategory.name();
 
-        LOGGER.info("Event time is {}, event code is {}, event category is {} and event counter is {}", dateTime, code,
+        log.info("Event time is {}, event code is {}, event category is {} and event counter is {}", dateTime, code,
                 eventLogCategoryName, eventCounter);
 
         // build a new EventDto with those values.
-        return new EventDto(dateTime, code.intValue(), eventCounter, eventLogCategoryName);
+        return new EventDto(dateTime, code.intValue(), eventCounter, eventLogCategoryName, null, null);
     }
 
-    private DateTime extractDateTime(final List<DataObject> eventData) throws ProtocolAdapterException {
+    private EventDto getPowerFailureEvent(List<DataObject> eventData, final EventLogCategoryDto eventLogCategory)
+            throws ProtocolAdapterException {
+        final DateTime endTime = this.extractDateTime(eventData);
+        final Long duration = this.extractEventDuration(eventData);
+        final String eventLogCategoryName = eventLogCategory.name();
+        final DateTime startTime = calculatePowerFailureStartTime(endTime, duration);
+
+        log.info("PowerFailureEvent startTime is {}, endTime is {}, event duration is {} and event category is {}",
```

Comment:
could be simplified as an event.toString method

Reply:
There is no event and the DTO doesn't have a toString. Are you suggesting to implement one in the DTO and log the toString before returning the object?